### PR TITLE
feat(util): atomWithDefault

### LINKED ISSUE #352: Writeable Atom with derived default value
I just recently went through a refactoring from recoil. Everything was pretty straightforward but one use case I couldn't build in jotai.

I have a list of countries I can fetch from the backend:

```js
const availableCountriesApi = atom<Array<Country>>(async () => fetchEntities("/core/countries"))
```

I built a defaultCountry atom based on this fetch

```js
const defaultCountry = atom<Country>(
  (get) => get(availableCountries).find(country => country.id === "US") || {id: "US", name: "United States"},
)
```

And now I have a `selectedCountry` atom which I want to initialize with the default country. But everything I tried resulted either in a type error or a readable only atom:

```js
const selectedCountryState = atom<Country>((get) => get(defaultCountry))
 
const [selectedCountry, setSelectedCountry] = useAtom(selectedCountryState)

setSelectedCountry({id: "DE", name: "Germany"}) // This expression is not callable, Type 'never' has no call signatures.
```

I think the async fetch is not super relevant, because when I change my defaultCoutry atom to this, I'll run in the same problem:

```js
const defaultCountry = atom<Country>({id: "US", name: "United States"})
```

So is there a way to build a writeable atom that I can initialize with a derived atom value?

### PR DESCRIPTION
**Pull request by dai-shi**
Close #352 

- [x] impl
- [x] tests
- [x] docs

### CODE REVIEW

File: src/utils/atomWithReset.ts [hunk 1]
```diff
@@ -6,7 +6,7 @@ export const RESET = Symbol()
 
 export function atomWithReset<Value>(initialValue: Value) {
   type Update = SetStateAction<Value> | typeof RESET
-  const anAtom: any = atom<Value, Update>(initialValue, (get, set, update) => {
+  const anAtom = atom<Value, Update>(initialValue, (get, set, update) => {
```
Comment (dai-shi) [inline src/utils/atomWithReset.ts:9]: just found we don't need any here.

File: docs/api/utils.md [hunk 2]
```diff
@@ -180,6 +180,22 @@ const countReducerAtom = atomWithReducer(0, countReducer)
...
+Ref: https://github.com/pmndrs/jotai/issues/352
+
+This is a function to create a primitive atom.
+Its default value can be specified with a read function instead of a static initial value.
+
+```js
+import { atomWithDefault } from 'jotai/utils'
+
+const count1Atom = atom(1)
+const count2Atom = atomWithDefault((get) => get(count1Atom) * 2)
```
Comment (Thisen) [inline docs/api/utils.md:194]: Question:
Is it possible to add a variant, where you can add a default value in atom? Could be useful, if you want to base the default value on props.

File: src/utils/atomWithDefault.ts [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+      if (get(overwrittenAtom)) {
+        return get(anAtom)
+      }
+      return getDefault(get)
+    },
+    (get, set, update) => {
+      set(overwrittenAtom, true)
+      set(
+        anAtom,
+        typeof update === 'function'
```
Comment (Thisen) [inline src/utils/atomWithDefault.ts:19]: Can it be a promise with this impl?

File: tests/utils/atomWithDefault.test.tsx [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+  fireEvent.click(getByText('button1'))
+  await findByText('count1: 3, count2: 5')
+})
+
+it('simple async get default', async () => {
+  const count1Atom = atom(1)
+  const count2Atom = atomWithDefault(async (get) => {
+    await new Promise((r) => setTimeout(r, 10))
+    return get(count1Atom) * 2
```
Comment (Thisen) [inline tests/utils/atomWithDefault.test.tsx:48]: Maybe a test missing with async `set`

File: docs/api/utils.md (hunk 2 above)
Comment (dai-shi) [reply docs/api/utils.md:194]: I'm not sure if I get "add a default value in atom".
Could you write some rough idea/code?

File: src/utils/atomWithDefault.ts (hunk 3 above)
Comment (dai-shi) [reply src/utils/atomWithDefault.ts:19]: I don't think so. The created atom is just like a normal primitive atom.

File: tests/utils/atomWithDefault.test.tsx (hunk 4 above)
Comment (dai-shi) [reply tests/utils/atomWithDefault.test.tsx:48]: Hadn't thought about it. Would it be possible for you to add such a test?

File: docs/api/utils.md (hunk 2 above)
Comment (Thisen) [reply docs/api/utils.md:194]:
```
const DogInfo: FC = (props) => {
  const dogInfo = useAtom(dogAtom(props.dogName))
  // ...
}
```

File: tests/utils/atomWithDefault.test.tsx (hunk 4 above)
Comment (Thisen) [reply tests/utils/atomWithDefault.test.tsx:48]: Sure, I'll do it

File: src/utils/atomWithDefault.ts (hunk 3 above)
Comment (Thisen) [reply src/utils/atomWithDefault.ts:19]: Is that acceptable?

**Comment by aulneau** [comment]
This is a great helper, I was wondering if it makes sense to have `atomFamilyWithDefault`, too?

File: src/utils/atomWithDefault.ts (hunk 3 above)
Comment (dai-shi) [reply src/utils/atomWithDefault.ts:19]: I mean it should be the same as the normal `atom(...)`. As far as I remember, if we set atom with a promise, it works for some case, but had a limitation or pitfall. At least the type doesn't allow it, so we are not currently supporting such a pattern.

File: docs/api/utils.md (hunk 2 above)
Comment (dai-shi) [reply docs/api/utils.md:194]: Ah, I want to support this pattern with this:
```js
const DogInfo = ({ dogName }) => {
  const dogAtom = useMemo(() => atom(dogName), [dogName])
  const dogInfo = useAtom(dogAtom)
  // ...
}
```

Oh, I see you want to give the default that can be overwritten.
```js
const DogInfo = ({ dogName }) => {
  const dogAtom = useRef(atom(dogName)).current
  const dogInfo = useAtom(dogAtom)
  // ...
}
```
I'm not a big fan of this useRef pattern.

```js
const DogInfo = ({ dogName }) => {
  const dogAtom = useMemo(() => atom(null), [])
  const dogInfo = useAtom(dogAtom) || dogName
  // ...
}
```

There might be some better ideas. I feel like using atomWithDefault for props is a misuse.

**Comment by dai-shi** [comment]
> This is a great helper, I was wondering if it makes sense to have `atomFamilyWithDefault`, too?

tbh, i'm not very happy with it. we have atomWithRest too. we may have more helpers in the future.
I now think it's better if atomFamily takes atom creator instead of read/write.
```js
const myFamily = atomFamily(param => atom(...))
// then this is possible
const myFamily = atomFamily(param => atomWithDefault(...))
```

what do you think? worth a breaking change before v1?

**Comment by aulneau** [comment]
> what do you think? worth a breaking change before v1?

I do think so. 

I have noticed that for any of the utility atoms, I've been needing to wrap them with other kinds. such as wishing i could do `atomFamilyWithQuery`. I could also see a need for something like this `atomFamilyWithDefaultWithQuery` lol (naming just for functional example)

**Comment by dai-shi** [comment]
Okay, the type would be much better too. Let's do that.
Would you or anyone like to work on it? I'm happy to take it if nobody raises a hand.
Maybe we want to create a dedicated issue at first.

**Comment by dai-shi** [comment]
> Maybe we want to create a dedicated issue at first.

Done: #385

File: tests/utils/atomWithDefault.test.tsx (hunk 4 above)
Comment (dai-shi) [reply tests/utils/atomWithDefault.test.tsx:89]: I first thought it's a neat idea, but then this doesn't work... unfortunately. async just works more or less accidentally. the `write` may not usually write self (though technically possible except for types), so this is counter-idiomatic too.
Thanks for your work anyway. I got an idea to improve the impl. (I did something weird, to try something else.)

**Review by Thisen — APPROVED** [review]

**Comment by dai-shi** [comment]
Thanks for the review. I will merge this later with others.

**Pull request merged**